qiskit_machine_learning/neural_networks/opflow_qnn.py

```diff
@@ -0,0 +1,136 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""An Opflow Quantum Neural Network that allows to use a parametrized opflow object as a
+neural network."""
+
+from copy import deepcopy
+from typing import List, Optional, Union, Tuple, Dict
+
+import numpy as np
+from qiskit.circuit import Parameter
+from qiskit.opflow import Gradient, CircuitSampler, ListOp, OperatorBase, ExpectationBase
+from qiskit.providers import BaseBackend, Backend
+from qiskit.utils import QuantumInstance
+from qiskit.utils.backend_utils import is_aer_provider
+
+from .neural_network import NeuralNetwork
+from .. import QiskitMachineLearningError
+
+
+class OpflowQNN(NeuralNetwork):
+    """Opflow Quantum Neural Network."""
+
+    def __init__(self, operator: OperatorBase,
+                 input_params: Optional[List[Parameter]] = None,
+                 weight_params: Optional[List[Parameter]] = None,
+                 exp_val: Optional[ExpectationBase] = None,
+                 gradient: Optional[Gradient] = None,
+                 quantum_instance: Optional[Union[QuantumInstance, BaseBackend, Backend]] = None):
+        """Initializes the Opflow Quantum Neural Network.
+
+        Args:
+            operator: The parametrized operator that represents the neural network.
+            input_params: The operator parameters that correspond to the input of the network.
+            weight_params: The operator parameters that correspond to the trainable weights.
+            exp_val: The Expected Value converter to be used for the operator.
+            gradient: The Gradient converter to be used for the operator's backward pass.
+            quantum_instance: The quantum instance to evaluate the network.
+        """
+        self.operator = operator
+        self.input_params = list(input_params or [])
+        self.weight_params = list(weight_params or [])
+        self.exp_val = exp_val  # TODO: currently not used by Gradient!
+        self.gradient = gradient or Gradient()
+
+        if isinstance(quantum_instance, (BaseBackend, Backend)):
+            quantum_instance = QuantumInstance(quantum_instance)
+
+        if quantum_instance:
+            self.quantum_instance = quantum_instance
+            self.circuit_sampler = CircuitSampler(
+                self.quantum_instance,
+                param_qobj=is_aer_provider(self.quantum_instance.backend)
+            )
+            # TODO: replace by extended caching in circuit sampler after merged: "caching='all'"
```

Comment:
FYI This has been merged already Qiskit/qiskit-terra#5934

Reply:
Yes, now make use of `caching="all"`.